### ISSUE TITLE
Separate windowing from display code

### ISSUE THREAD

**Issue by rdb** [panda3d/panda3d#1128]
## Problem
Currently, windowing code is part of the display module.  There is the GraphicsWindow base class, which is inherited by x11GraphicsWindow, which is inherited by both glxGraphicsWindow and eglGraphicsWindow.

This is making it difficult to switch Panda to use OpenGL without X11 (using EGL instead of GLX) without recompiling (see #1086, #557).  Creating a separate display module using EGL is an acceptable workaround for now, but EGL could actually be used with or without X11.  This would also become untenable in the longer term when we add support for Wayland and Vulkan, since the ability to mix-and-match windowing and display API would create a combinatoric explosion of display modules:
* OpenGL + X11 (via GLX)
* OpenGL + Wayland
* OpenGL headless (via EGL)
* OpenGL ES + X11
* OpenGL ES + Wayland
* Vulkan + X11
* Vulkan + Wayland
* Any of the above but without windowing module, for headless applications.

## Proposed Solution
The first step to solving this, in my opinion, is to split out the windowing code from the display module.  In the short term, or for platforms where we only support one windowing system, we could move it to the core library.  Eventually, we would have something like a plug-in "platform module" or "app module", such as for X11 (or for SDL2), Windows, macOS, Oculus Rift, etc.  (I'm reluctant to use the term "window" since it is specific to desktop games.)

This leaves the question of where we put the EGL/GLX code.  It would be convenient if we could use EGL across the board on Linux, but I'm not sure that we can consistently rely on EGL being available.  I think the display module will need to be able to dynamically load the EGL libraries, and dynamically dispatch to a GLX-based or EGL-based implementation depending on what kind of host window is provided and what is available on the system.  An alternative is to reimplement the windowing code using SDL instead, which would obviate the need for us to deal with this.

I think this would require that the GraphicsWindow class is split into a class handling the actual window on screen, and a class handling the window's framebuffer and display-related matters.  Doing this split in a way that doesn't break backward compatibility will be tricky.

In order not to break the most basic API interactions people have with `base.win` (such as `requestProperties`), I think `base.win` should continue to represent the "OS window", but it would no longer inherit from GraphicsOutput.  The underlying GraphicsOutput object would instead be created when the window is opened, and stored as eg. a `framebuffer` attribute.  This would break the display-oriented API calls such as `makeDisplayRegion()`, though we can perhaps add these as deprecated forwarders for now.

More discussion and design work on an API is needed.

**Comment by rdb** [comment]
Some more thoughts on this.

EGL / GLX linking is a non-issue.  We can load libEGL with dlopen (it's just a handful of functions), and GLX doesn't have its own library, its symbols are already part of libGL.so.  That also means that every OpenGL application on Linux already implicitly depends on X11—this is unfortunate, but the simple reality of OpenGL on Linux, and not a problem we can solve anyway.  So, the answer to the question of "where does this code live" is "in libpandagl.so".

If we have a Wayland implementation, due to the design of Wayland, it can link to libwayland-client statically (except perhaps on platforms where it is widely available) to avoid a runtime dependency, so this is also a non-issue.

Similar to having a "OS Window", I also think we need an "OS App" class.  This takes the windowing-specific state out of GraphicsPipe and simultaneously solves the problem that most OSes have an app-wide event loop, as opposed to a per-window event loop.  Furthermore, it gives access to the OS application abstraction that some operating systems have, eg. allowing setting an application icon on macOS or managing app-wide configuration.

This may be out of scope for now, but we should also consider having this class also be in charge of input propagation and stepping the task manager, taking over this duty from ShowBase.  The reason for this is that some platforms (most strongly web, but others too) prefer to be in charge of deciding when rendering occurs.

**Comment by Moguri** [comment]
I like the idea of making a more app-focused instead of window-focused workflow. I find it difficult to do these refactoring changes (splitting up displays and/or OS app) without breaking backwards compatibility, which means it's probably better suited for a Panda3D 2.0 clean up.

In the meantime, an sdldisplay (with derived classes for GL, GL ES?, D3D9, and Vulkan) would reduce class explosion and give us Wayland support and a path forward for Vulkan. We should keep the existing displays for backwards compatibility and as a fallback if something isn't working right with the SDL versions.

For Panda 2.0 we can evaluate how SDL is doing and if we can prune any of our displays. At this point we could also determine if we gain much by splitting window and display code. If SDL suits our needs and becomes the only window code, then splitting doesn't gain much. If we find we prefer to keep our own per-platform windowing code, then we'll want to fix the combinatorial class explosion.

One thing that throws a bit of a wrench into this plan (timing-wise) is that there is a new major version of SDL (SDL 3) in the works. I don't know which version (2 vs 3) will make sense to target or even what kind of an effort a 2 to 3 migration will be.
